jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -350,119 +351,363 @@ public void abort(Throwable failure)
+                        return;
+                    }
+                }
             }
-            if (LOG.isDebugEnabled())
-                LOG.debug(String.format("HTTP2 Response #%d/%h %s %s", stream.getId(), stream.getSession(),
-                    commit ? "commit" : "flush", callback == null ? "ignored" : "failed"), failure);
-            if (callback != null)
-                callback.failed(failure);
+            fail(callback, commit, failure);
         }
 
-        private boolean onIdleTimeout(Throwable failure)
+        private boolean idleTimeout(Throwable failure)
         {
-            boolean result;
-            Callback callback = null;
+            Callback callback;
+            boolean timeout;
             synchronized (this)
             {
-                // Ignore idle timeouts if not writing,
-                // as the application may be suspended.
-                result = state == State.WRITING;
-                if (result)
+                switch (_state)
                 {
-                    this.state = State.TIMEOUT;
-                    callback = this.callback;
-                    this.callback = null;
-                    this.failure = failure;
+                    case PENDING:
+                    {
+                        // The send was started but idle timed out, fail it.
+                        _state = State.FAILED;
+                        _failure = failure;
+                        callback = _callback;
+                        timeout = true;
+                        break;
+                    }
+                    case IDLE:
+                        // The application may be suspended, ignore the idle timeout.
+                    case SENDING:
+                        // A send has been started at the same time of an idle timeout;
+                        // Ignore the idle timeout and let the write continue normally.
+                    case SUCCEEDING:
+                    case FAILING:
+                        // An idle timeout during these transient states is ignored.
+                    case FAILED:
+                        // Already failed, ignore the idle timeout.
+                    {
+                        callback = null;
+                        timeout = false;
+                        break;
+                    }
+                    default:
+                    {
+                        // Should not happen, but just in case.
+                        _state = State.FAILED;
+                        _failure = new IllegalStateException("Invalid transport state: " + _state, failure);
+                        callback = _callback;
+                        if (callback == null)
+                            callback = Callback.NOOP;
+                        timeout = true;
+                        failure = _failure;
+                        break;
+                    }
                 }
             }
+            idleTimeout(callback, timeout, failure);
```

Comment:
just inline the idleTimeout method here.

Reply:
Nah, it nicely complements the succeed() and fail() methods, with logging callback action.